FindESMF.cmake

```diff
@@ -0,0 +1,105 @@
+function(append_globbed_directories VAR)
+	cmake_parse_arguments(ARGS
+		"" 
+		""
+		"PATTERNS;PATHS" 
+		${ARGN}
+	)
+	set(MATCHED_LIST "")
+	foreach(PREFIX ${ARGS_PATHS})
+        foreach(PATTERN ${ARGS_PATTERNS})
+            if(IS_ABSOLUTE ${PREFIX})
+                file(GLOB MATCHED ${PREFIX}/${PATTERN})
+            else()
+                file(GLOB MATCHED ${CMAKE_BINARY_DIR}/${PREFIX}/${PATTERN})
+            endif()
+            foreach(MATCHED_FILE ${MATCHED})
+                get_filename_component(MATCHED_DIR ${MATCHED_FILE} DIRECTORY)
+                list(APPEND MATCHED_LIST ${MATCHED_DIR})
+            endforeach()
+		endforeach()
+    endforeach()
+    if("${MATCHED_LIST}")
+        list(REMOVE_DUPLICATES MATCHED_LIST)
+	endif()
+	list(APPEND ${VAR} ${MATCHED_LIST})
+	set(${VAR} ${${VAR}} PARENT_SCOPE)
+endfunction()
+
+# Add globbed directories to CMAKE_PREFIX_PATH for upcoming find_paths
+append_globbed_directories(CMAKE_PREFIX_PATH
+	PATTERNS
+		mod/mod*/*.*.*.*.*/esmf.mod
+		lib/lib*/*.*.*.*.*/libesmf.a
+	PATHS
+		${CMAKE_PREFIX_PATH}
+)
+
+# Find the installed ESMF files
+find_path(ESMF_HEADERS_DIR
+	ESMC.h
+	DOC "The path to the directory containing \"ESMC.h\"."
+	PATH_SUFFIXES "include"
+)
+
+find_path(ESMF_MOD_DIR
+	esmf.mod
+	DOC "The path to the directory containing \"esmf.mod\"."
+	PATH_SUFFIXES "mod" "include"
+)
+
+find_library(ESMF_LIBRARY
+	libesmf.a
+	DOC "The path to the directory containing \"libesmf.a\"."
```

Comment:
Note that before we were using `libesmf_fullylinked.so` and not the static library (when on Linux). Is there a reason you went full static?

Reply:
Only because that's what the original GCHP build system linked. I could add a cached variable like `ESMF_FIND_LIBRARY_NAMES` and initialize it to whatever GEOS-ESM's preference is. Should it be `esmf` on OSX and `esmf_fullylinked` everywhere else?

Reply:
Actually, this might be more interesting than I think. I'm trying to get GEOSgcm working with CircleCI after our upgrade to MAPL 2.0 and for reasons I cannot figure out, it will not link to `esmf_fullylinked`. 

I'm currently trying to figure out if I can do a `sed` hack in the CI to force the static library and see if that works. If so, I might have to add a fix to `master` here where I can do something like `-DESMF_STATIC=YES`.

Reply:
(Note: even ESMF *itself* is sometimes having issues linking to its own `esmf_fullylinked` when it builds its own apps during a Baselibs build. Something odd is happening.)

Reply:
This is fixed in 7805c686064a05d47f7fa2c0d7a5a982a09eac52.

A cache variable called `FIND_ESMF_LIBRARY_SEARCH_NAMES` stores the library names that are searched for in the `find_library(ESMF_LIBRARY ...) call. I set the default value to "esmf;esmf_fullylinked" but this can be overwritten by making your own cache entry before calling `include(esma)`. For example, to link `libesmf_fullylinked.a` you would do

```cmake
set(FIND_ESMF_LIBRARY_SEARCH_NAMES "libesmf_fullylinked.a" CACHE INTERNAL "I want to use libesmf_fullylinked.a") # this line before include(esma)
include(esma)
```